### ISSUE TITLE
How can I create a multitone?

### ISSUE THREAD

**Issue by glagola** [PHP-DI/PHP-DI#593]
Hi, I hit some restrictions during using php-di.

**Use case**
I'm using identity map in my repositories to reduce amount of requests to the database for the web. But for the background worker scripts I need transparent/fake identity map implementation (data changes fast, and instances stored to the identity map became invalid instantly). So I created common interface for both implementations called `IdentityMapInterface`. For each repository I need to have individual identity map (not shared instance).

How can I define a multitone (opposite to singleton) instance of the identity map interface in container?

**My current workaround**:
```
$containerBuilder->addDefinitions([
    // ...
   FileRepositoryInterface::class => DI\autowire(FileRepository::class)
        ->constructorParameter('identityMap', DI\create(IdentityMap::class)),
   
    FileMetadataRepository::class => DI\autowire()
        ->constructorParameter('identityMap', DI\create(IdentityMap::class)),
    // ...
]);
```
The workaround allows to create separate/individual `IdentityMap` for each repository.

Is there any way to do it better?

**I'm looking something like this**:
```
$containerBuilder->addDefinitions([
    // ...
   IdentityMapInterface::class => DI\create(IdentityMap::class)->multitone(),
    // ...
]);
```
It will allow me to reduce the code duplication of injection identityMap as constructor parameter (like I did in the workaround), autowiring of the container will resolve everything by itself.

**Comment by mnapoli** [comment]
I see, is there a reason why each repository cannot create the `IdentityMap` itself with `$this->identityMap = new IdentityMap` in the constructor?

That would avoid having to go through PHP-DI for that. Because with this:

```php
$containerBuilder->addDefinitions([
    // ...
   FileRepositoryInterface::class => DI\autowire(FileRepository::class)
        ->constructorParameter('identityMap', DI\create(IdentityMap::class)),
   
    FileMetadataRepository::class => DI\autowire()
        ->constructorParameter('identityMap', DI\create(IdentityMap::class)),
    // ...
]);
```

each instance of `IdentityMap` is created and passed to each repository and PHP-DI does nothing more then…

Or maybe you cannot use `new` because `IdentityMap` takes dependencies? In that case you might want to use `make()` (see http://php-di.org/doc/container.html#make). You can inject the container but still decouple from it using the `DI\FactoryInterface` interface.

Or if you don't want to use `DI\FactoryInterface` I would suggest you create your own `IdentityMapFactory` and inject that factory in every repository. Then you can do `$this->identityMap = $factory->create()` in each constructor.

**Comment by glagola** [comment]
> is there a reason why each repository cannot create the IdentityMap itself with $this->identityMap = new IdentityMap in the constructor?

Yes, according to the use case, repository should not know if IdentityMap is fake or not, that is why he can't create it by itself.

> I would suggest you create your own IdentityMapFactory and inject that factory in every repository. Then you can do $this->identityMap = $factory->create() in each constructor.

That is much better idea, thanks! I'll create a interface for Identity map factory, and will write two implementations for it (for fake one and for real one).

Btw, It would be great to write somewhere directly in documentation, that it is not possible to create multiple instances of dependency via the DI container. I had to read all documentation and try few times before I realize that it is not possible.

**Comment by mnapoli** [comment]
Right, there is this page: http://php-di.org/doc/scopes.html But it doesn't appear in the menus anymore (since it has been removed in PHP-DI 6). I kept the page up and I updated it in case it was referenced somewhere on the web.

I'll keep this issue open for adding a piece to the documentation.

**Comment by glagola** [comment]
>  there is this page: http://php-di.org/doc/scopes.html

Yes, I found the page via google, it was a key point in my conclusion that it is not possible to implement.
